[FLINK-25433][runtime] Adds retry mechanism to DefaultResourceCleaner

XComp opened the pull request:
## What is the purpose of the change

We want to make the cleanup try to run infinitely until the cleanup is resolved.

## Brief change log

* Adds `FutureUtils.retryWithDelay` in `DefaultResourceCleaner`

## Verifying this change

* `DefaultResourceCleanerTest` was extended to cover the retry mechanism

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java:
```diff
@@ -21,42 +21,64 @@
+ */
 public class DefaultResourceCleaner<T> implements ResourceCleaner {
 
+    private static final RetryStrategy DEFAULT_RETRY_STRATEGY =
```
I wasn't sure about making this properties configurable. Initially, we said we don't need to do that as far as I remember. But thinking about it: At least the initial delay might be interesting for the user?

metaswirl commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleanerTest.java:
```diff
@@ -181,33 +212,97 @@ public void testMediumPriorityCleanupBlocksAllLowerPrioritizedCleanups() {
 
-        public CompletableFuture<Void> cleanup(JobID jobId, Executor executor) {
+        public CompletableFuture<Void> apply(JobID jobId, Executor executor) {
             Preconditions.checkState(this.jobId == null);
             this.jobId = jobId;
```
Would it make sense to wrap this in a synchronized, to avoid a race condition?

XComp replied:
Yes, the contract is that the method's body is run from within the main thread (see [DefaultResourceCleaner:127](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java#L127). Hence, all the synchronization happens through the main thread task queue. The test implementation does not add any concurrency.

metaswirl commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleanerTest.java:
```diff
@@ -181,33 +212,97 @@ public void testMediumPriorityCleanupBlocksAllLowerPrioritizedCleanups() {
+
+        final CompletableFuture<Void> compositeCleanupResult =
+                createTestInstanceBuilder(new FixedRetryStrategy(1, Duration.ZERO))
+                        .withPrioritizedCleanup(cleanupWithRetry)
```
How about we add a second prio cleanup here, to show that the second cleanup is executed a single time, even if the first prio cleanup retries.

XComp replied:
I think that makes sense since the prio tasks are run in the same for loop. That way we can ensure that the retry really only done for the individual cleanup task. 👍

metaswirl reviewed (changes requested):
How about we add a test that covers that the cleanup ultimately fails after exceeding retries.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleanerTest.java:
```diff
@@ -181,33 +212,97 @@ public void testMediumPriorityCleanupBlocksAllLowerPrioritizedCleanups() {
+
+        final CompletableFuture<Void> compositeCleanupResult =
+                createTestInstanceBuilder(new FixedRetryStrategy(1, Duration.ZERO))
+                        .withPrioritizedCleanup(cleanupWithRetry)
```
Yes, the prioritized cleanup is necessary for the global cleanup. We have to clean the `JobManagerRunnerRegistry` before cleaning the `HighAvailabilityServices`. There is a watcher implementation in the k8s leader election that complains/throws an Exception if we clean ConfigMaps while not being the leader. That could happen if we run the `HighAvailabilityServices` cleanup before stopping the leader election of the `JobMaster` (which happens in the JobMaster close business logic).

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleanerTest.java:
```diff
@@ -19,50 +19,52 @@
 /** {@code DefaultResourceCleanerTest} tests {@link DefaultResourceCleaner}. */
 public class DefaultResourceCleanerTest {
 
+    private static final Duration TIMEOUT = Duration.ofMillis(100);
```
true

XComp replied:
Looks like I was wrong ([build failure](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=30875&view=logs&j=0da23115-68bb-5dcd-192c-bd4c8adebde1&t=24c3384f-1bcb-57b3-224f-51bf973bbee8&l=9374)): We need the timeout. Although, I'm not sure, yet, why. My understanding is due to the test using the `Executors.directExecutors()`, the execution should be handled in a non-concurrent fashion. That makes me believe that the `CompletableFuture` should have finished by the time we reach the asserts 🤔

XComp replied:
Tests that include retries are subject to failure if we're not using a timeout because the retry mechanism utilizes the `ComponentMainThreadExecutor.schedule` method which runs the retry on a separate thread through the `DirectScheduledExecutorService`.

XComp commented:
> How about we add a test that covers that the cleanup ultimately fails after exceeding retries.

That's a good idea. But I realized that this is already covered by the `testConcurrentCleanupWithException*` tests. Therefore, I left it without the extra test.

XComp commented:
I addressed the comments, and in addition to that, added a `*ITCase` (after renaming the `DispatcherFailOverITCase` into `DispatcherCleanupITCase`) covering the retry logic in a bigger context.

metaswirl reviewed (changes requested):
Looks good. Some minor comments.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java:
```diff
@@ -114,32 +137,37 @@ private DefaultResourceCleaner(
                                         .collect(Collectors.toList())));
     }
+
+    private CompletableFuture<Void> withRetry(JobID jobId, T cleanup) {
```
We're always passing the cleanup as a parameter already: `.map(cleanup -> withRetry(jobId, cleanup))`. I feel like it doesn't add any value considering that the operation is passed via the `cleanup` parameter and it's a private method, anyway

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherCleanupITCase.java:
```diff
@@ -110,70 +113,117 @@ public void tearDown() {
+                        .setFatalErrorHandler(
+                                throwable -> {
+                                    final Optional<Throwable> maybeError =
+                                            ExceptionUtils.findThrowable(
+                                                    throwable, temporaryError::equals);
+                                    if (maybeError.isPresent()
+                                            && temporaryError.equals(maybeError.get())) {
```
Good catch 👍

XComp commented:
Thanks for the reviews @dmvk and @metaswirl . I squashed the changes and rebased the branch. I will merge the PR after the final AzureCI run succeeded.

XComp commented:
Looks like rebasing wasn't a good idea. There are failures on `master` covered by FLINK-26065. Additionally, the e2e tests stopped due to some problem that looks like an infra issue

XComp commented:
@flinkbot run azure
